api/external.go

```diff
@@ -90,39 +139,93 @@ func (a *API) ExternalProviderCallback(w http.ResponseWriter, r *http.Request) e
 func (a *API) internalExternalProviderCallback(w http.ResponseWriter, r *http.Request) error {
 	ctx := r.Context()
 	config := a.getConfig(ctx)
-	instanceID := getInstanceID(ctx)
 
-	providerType := getExternalProviderType(ctx)
+	// handle callback using session data
+	providerType, err := gothic.GetProviderName(r)
+	if err != nil {
+		return err
+	}
+
 	var userData *provider.UserProvidedData
 	var providerToken string
-	if providerType == "saml" {
-		samlUserData, err := a.samlCallback(r, ctx)
+
+	gothProvider, err := goth.GetProvider(providerType)
+	if err != nil {
+		return err
+	}
+	value, err := gothic.GetFromSession(providerType, r)
+	if err != nil {
+		return err
+	}
+	sess, err := unmarshalSession(value, gothProvider)
+	if err != nil {
+		return err
+	}
+	// validate state to ensure that the state token param from the original
+	// AuthURL matches the one included in the current callback req
+	rawAuthURL, err := sess.Session.GetAuthURL()
+	if err != nil {
+		return err
+	}
+	authURL, err := url.Parse(rawAuthURL)
+	if err != nil {
+		return err
+	}
+	reqState := gothic.GetState(r)
+	originalState := authURL.Query().Get("state")
+	if originalState != "" && (originalState != reqState) {
+		return errors.New("state token mismatch")
+	}
+	gothUser, err := gothProvider.FetchUser(sess.Session)
+	if err != nil {
+		params := r.URL.Query()
+		if params.Encode() == "" && r.Method == "POST" {
+			r.ParseForm()
+			params = r.Form
+		}
+		_, err = sess.Session.Authorize(gothProvider, params)
 		if err != nil {
 			return err
 		}
-		userData = samlUserData
-	} else {
-		oAuthResponseData, err := a.oAuthCallback(ctx, r, providerType)
+		gothUser, err = gothProvider.FetchUser(sess.Session)
 		if err != nil {
 			return err
 		}
-		userData = oAuthResponseData.userData
-		providerToken = oAuthResponseData.token
+	}
+
+	providerToken = gothUser.AccessToken
+	userData = &provider.UserProvidedData{
+		Metadata: map[string]string{
+			"full_name":  gothUser.Name,
+			"avatar_url": gothUser.AvatarURL,
+		},
+	}
+
+	if gothUser.Email != "" {
+		userData.Emails = append(userData.Emails, provider.Email{
+			Email:    gothUser.Email,
+			Verified: true, // TODO: check with external provider if email is verified
+			Primary:  true, // TODO: check with external provider if email is primary email
```

Comment:
Some providers return information on whether the email has been verified and / or whether the email is the user's primary email in the account. Currently, I'm making the assumption that the email returned by the external provider is verified and is the primary email.